### PR TITLE
Lad metrics

### PR DESCRIPTION
**Pull request by nidhanda**
Added telegraf config parser logic and incorporated it into the LAD code for now. 

**Work still needed**
1) ME set up script and integration

**Can Ignore**
1) The telegraf config files  (Including them here temporarily only for reference, their format will change along with the script and **will not be part of the final code merge**.)

Will keep updating this PR once the further changes are added.

### CODE REVIEW

File: Diagnostic/test.settings [hunk 1]
```diff
@@ -0,0 +1 @@
...e=TRUE", "counter": "averagediskqueuelength", "counterSpecifier": "/builtin/disk/averagediskqueuelength", "type": "builtin", "unit": "Count", "sampleRate": "PT15S"}]}}, "sampleRateInSeconds": 15}}, "protectedSettings": "MIICmwYJKoZIhvcNAQcDoIICjDCCAogCAQAxggFpMIIBZQIBADBNMDkxNzA1BgoJkiaJk/IsZAEZFidXaW5kb3dzIEF6dXJlIENSUCBDZXJ0aWZpY2F0ZSBHZW5lcmF0b3ICEF2q8a+rPWOCTbrYYrJiNPAwDQYJKoZIhvcNAQEBBQAEggEAi7PjaBKLbeuqvBsKGB7FBYSw8Ikxo3567DtFuXy+LkKU/Ms4DQZnnij3ubo4/LYx3qNeo5quS/niY61Z3uS1s8J70EPY9KHfbJXnzdzD7dc2fkNzw1ZROplmMwepJSWQEc//aYe5g7ZsBK3aOYlO2Z30DbampoSVKlb99bsi1wQc4pKTtLCJig/Le0D997T5MWnDfKzQUUrWye0vYJe64k9siMNDrsbIkQOcjsXq9MBh4SmyznYJ3HQ6IPeahqoVe3XdogAK7stCnWMSLmtW4rQQXEjNiwYWShVxHv7COd315zbDIflMWBTKMNh55Tltk+5kkbZLtmWQsG63fWXNyDCCARQGCSqGSIb3DQEHATAUBggqhkiG9w0DBwQIBwy3u4+xh0qAgfAt+5PVrEZ8PqIbzZ5bCq9fJgpnccYDId86uACOlyXNRDAqREHL28Zyzw3pzQ4kVCieJew931Gc1D80Oo1EZspjzAScK+qIQn4SG6sSF51SaKMb+aABggrOficPVf1tVJ2vxHAXpXpW31+Dm9JGqxrbpAGQ9BN9yCYuwt6NI5mwrO52j5tiFx7sfliD6vgF0S1zy24CrcOmchwcqWnn4BicXYRMPMtmxCtDEdzXyz8E5tDoIRmPdxpkhaRfldkXdvlzC7oqMzGfvL+ivec6z1GZNG/eVsCwmWDGfN3e+A4PW04Lj/jBNm4daShHUCDPvrg=", "protectedSettingsCertThumbprint": "7125DD096E0770BC8B90A54445E55111ADD8C0EC"}}]}
```
Comment (hestolz) [inline Diagnostic/test.settings:1]: uhhHH might not wanna expose secrets here

Comment (nidhanda) [reply Diagnostic/test.settings]: Ah good catch. removed it in the latest commit.

File: Diagnostic/Utils/lad_logging_config.py [hunk 2]
```diff
@@ -161,6 +162,77 @@ def get_syslog_ng_config(self):
...
+        """
+        if not ladconfig:
+            return []
+        data = []
+        default_sample_rate = "15s"  #Lowest supported time interval
+        if "sampleRateInSeconds" in ladconfig and ladconfig["sampleRateInSeconds"] != "":
+            default_sample_rate = str(ladconfig["sampleRateInSeconds"]) + "s" #Example, converting 15 to 15s
+
+        if 'diagnosticMonitorConfiguration' in ladconfig and "performanceCounters" in ladconfig['diagnosticMonitorConfiguration']:
+            data =  ladconfig['diagnosticMonitorConfiguration']["performanceCounters"]
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/Utils/lad_logging_config.py:210]: Added strip() and lower() below in line # 222

File: Diagnostic/Utils/lad_logging_config.py [hunk 3]
```diff
@@ -201,6 +273,33 @@ def __generate_mdsd_syslog_config(self):
         return mxt.top_level_tmpl_for_logging_only.format(
             sources=mxt.per_source_tmpl.format(name=syslog_src_name), events=mdsd_event_source, eh_urls=syslog_eh_urls)
 
+    def get_mdsd_telegraf_config(self, namespaces):
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/Utils/lad_logging_config.py:280]: Updated the docstring

File: Diagnostic/diagnostic.py [hunk 4]
```diff
@@ -315,6 +395,51 @@ def main(command):
...
+                    if me_msi_token_expiry_epoch:
+                        currentTime = datetime.datetime.now()
+                        token_expiry_time = datetime.datetime.fromtimestamp(me_msi_token_expiry_epoch)
+                        if token_expiry_time - currentTime < datetime.timedelta(minutes=30):
+                            # The MSI Token will expire within 30 minutes. We need to refresh the token
+                            generate_token = True
+
+                    if generate_token:
+                        generate_token = False
+                        msi_token_generated, me_msi_token_expiry_epoch, log_messages = me_handler.generate_MSI_token()
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/diagnostic.py:431]: Retries added in generate_MSI_token() method. Also this loop will self retry every 30 seconds

File: Diagnostic/copy_conf.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    for item in perfconf:
+        counter = item["counter"]
+        if counter in name_map:
+            plugin = name_map[counter]["plugin"]
+            if plugin not in telegraf_conf:
+                telegraf_conf[plugin] = {"fields" : {}, "displayName": item["class"]}
+            else:
+                telegraf_conf[plugin]["fields"][name_map[counter]["field"]] = {}
+                telegraf_conf[plugin]["fields"][name_map[counter]["field"]]["displayName"] = item["annotation"][0]["displayName"]
+                telegraf_conf[plugin]["fields"][name_map[counter]["field"]]["interval"] = item["sampleRate"][2:].lower() #Example, converting PT15S tp 15s
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/copy_conf.py]: isodate isn't shipped with python by default. Might cause issues with customers with basic version of python installed. I'll write a custom utility function to parse the PT* duration syntax.

File: Diagnostic/copy_conf.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        }, 
+        'displayName': u'processor'
+    }
+}
+"""
+input_str = ""
+rename_str= ""
+for plugin in telegraf_conf:
+    min_interval = "60s"
+    input_str += "[[inputs." + plugin + "]]\n"
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/copy_conf.py]: Stringio isnt included in python by default. + operator in python does basic concatenation and has proven to be quite reliable so far. Not sure if it's very feasible to include/install a third party library on the cx's VM

File: Common/metrics_ext_utils/metrics_ext_handler.py [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+    amrurl = "https://management.azure.com/subscriptions/" + subscription_id + "?api-version=2014-04-01"
+    try:
+        req = urllib2.Request(amrurl, headers={'Content-Type':'application/json'})
+        res = urllib2.urlopen(req)
+    except Exception as e:
+        err_res = e.headers["WWW-Authenticate"]
+        for line in err_res.split(","):
+                if "Bearer authorization_uri" in line:
+                        data = line.split("=")
+                        aad_auth_url = data[1][1:-1] #Removing the quotes from the front and back
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Common/metrics_ext_utils/metrics_ext_handler.py]: Will add Unit Tests for these in the Test PR.

File: Common/metrics_ext_utils/metrics_ext_handler.py [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+
+        else:
+            # No previous binary exist, simply copy it and make it executable
+            copyfile(me_bin_local_path, metrics_ext_bin)
+            os.chmod(metrics_ext_bin,stat.S_IXOTH)
+    else:
+        raise Exception("Unable to copy MetricsExtension Binary, could not find file at the location {0} . Failed to setup ME.".format(me_bin_local_path))
+        return False
+
+    me_influx_port = "8139"
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Common/metrics_ext_utils/metrics_ext_handler.py]: Modified them to use "13459" for LAD and  "17659" for AMA

File: Diagnostic/diagnostic.py [hunk 9]
```diff
@@ -277,6 +294,20 @@ def main(command):
             else:
                 stop_mdsd()
             oms.tear_down_omsagent_for_lad(RunGetOutput, False)
+
+            #Stop the telegraf and ME services
+            tel_out, tel_msg = telhandler.stop_telegraf_service(True)
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/diagnostic.py]: Done

File: Diagnostic/lad_config_all.py [hunk 10]
```diff
@@ -79,13 +81,18 @@ def __init__(self, ext_settings, ext_dir, waagent_dir, deployment_id,
         self._encrypt_secret = encrypt_string
         self._logger_log = logger_log
         self._logger_error = logger_error
+        self._telegraf_me_url = "udp://127.0.0.1:8139"
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/lad_config_all.py]: Added this in the metrics_constants file

File: Diagnostic/lad_config_all.py [hunk 11]
```diff
@@ -222,63 +212,63 @@ def _handle_alternate_sinks(self, interval, sinks, source):
...
-                if sink_type != 'JsonBlob' and sink_type != 'EventHub':
-                    raise LadPerfCfgConfigException('Sink type "{0}" (for sink name="{1}") is not supported'
-                                                    .format(sink_type, sink_name))
-                self._add_element_from_element('Events/OMI', generate_omi_query_xml_elem(omi_query, (sink_name, sink_type)))
-                if sink_type == 'EventHub':
-                    if 'sasURL' not in sink:
-                        raise LadPerfCfgConfigException('No sasURL specified for an EventHub sink (name="{0}")'
-                                                        .format(sink_name))
-                    self._add_streaming_annotation(sink_name, sink['sasURL'])
+    # def _update_raw_omi_events_settings(self, omi_queries):
```
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/lad_config_all.py]: Done

File: Diagnostic/copy_conf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nidhanda) [reply Diagnostic/copy_conf.py]: Added a custom utility to convert ISO 8601 time duration format into seconds

**Pull request merged**